exchanges/ftx/ftx_wrapper.go

```diff
@@ -0,0 +1,833 @@
+		return nil
+	}
+
+	err := f.SetupDefaults(exch)
+	if err != nil {
+		return err
+	}
+
+	err = f.Websocket.Setup(
+		&wshandler.WebsocketSetup{
+			Enabled:                          exch.Features.Enabled.Websocket,
+			Verbose:                          exch.Verbose,
+			AuthenticatedWebsocketAPISupport: exch.API.AuthenticatedWebsocketSupport,
+			WebsocketTimeout:                 exch.WebsocketTrafficTimeout,
+			DefaultURL:                       ftxWSURL,
+			ExchangeName:                     exch.Name,
+			RunningURL:                       exch.API.Endpoints.WebsocketURL,
+			Connector:                        f.WsConnect,
+			Subscriber:                       f.Subscribe,
+			UnSubscriber:                     f.Unsubscribe,
+			Features:                         &f.Features.Supports.WebsocketCapabilities,
+		})
+	if err != nil {
+		return err
+	}
+
+	f.WebsocketConn = &wshandler.WebsocketConnection{
+		ExchangeName:         f.Name,
+		URL:                  f.Websocket.GetWebsocketURL(),
+		ProxyURL:             f.Websocket.GetProxyAddress(),
+		Verbose:              f.Verbose,
+		ResponseCheckTimeout: exch.WebsocketResponseCheckTimeout,
+		ResponseMaxLimit:     exch.WebsocketResponseMaxLimit,
+	}
+
+	// NOTE: PLEASE ENSURE YOU SET THE ORDERBOOK BUFFER SETTINGS CORRECTLY
+	f.Websocket.Orderbook.Setup(
+		exch.WebsocketOrderbookBufferLimit,
+		false,
+		false,
+		false,
+		false,
+		exch.Name)
+	return nil
+}
+
+// Start starts the FTX go routine
+func (f *FTX) Start(wg *sync.WaitGroup) {
+	wg.Add(1)
+	go func() {
+		f.Run()
+		wg.Done()
+	}()
+}
+
+// Run implements the FTX wrapper
+func (f *FTX) Run() {
+	if f.Verbose {
+		log.Debugf(log.ExchangeSys,
+			"%s Websocket: %s.",
+			f.Name,
+			common.IsEnabled(f.Websocket.IsEnabled()))
+		f.PrintEnabledPairs()
+	}
+
+	if !f.GetEnabledFeatures().AutoPairUpdates {
+		return
+	}
+
+	err := f.UpdateTradablePairs(false)
+	if err != nil {
+		log.Errorf(log.ExchangeSys,
+			"%s failed to update tradable pairs. Err: %s",
+			f.Name,
+			err)
+	}
+}
+
+// FetchTradablePairs returns a list of the exchanges tradable pairs
+func (f *FTX) FetchTradablePairs(a asset.Item) ([]string, error) {
+	if a != asset.Spot && a != asset.Futures {
+		return nil, fmt.Errorf("asset type of %s is not supported by %s", a, f.Name)
+	}
+	markets, err := f.GetMarkets()
+	if err != nil {
+		return nil, err
+	}
+	var pairs []string
+	switch a {
+	case asset.Spot:
+		for x := range markets.Result {
+			if markets.Result[x].MarketType == spotString {
+				pairs = append(pairs, markets.Result[x].Name)
+			}
+		}
+	case asset.Futures:
+		for x := range markets.Result {
+			if markets.Result[x].MarketType == futuresString {
+				pairs = append(pairs, markets.Result[x].Name)
+			}
+		}
+	}
+	return pairs, nil
+}
+
+// UpdateTradablePairs updates the exchanges available pairs and stores
+// them in the exchanges config
+func (f *FTX) UpdateTradablePairs(forceUpdate bool) error {
+	pairs, err := f.FetchTradablePairs(asset.Spot)
```

Comment:
Instead of hardcoding asset.Spot/asset.Futures, it's best to enumerate the supported asset types (`for for x := range b.CurrencyPairs.AssetTypes`) and call FetchTradablePairs with the asset type